File: DroidsAPI/src/main/java/com/github/rochedo098/droidsapi/api/events/ModuleEventHandler.java
package com.github.rochedo098.droidsapi.api.events;

import net.minecraft.entity.Entity;
import org.jetbrains.annotations.ApiStatus;

import java.util.ArrayList;
import java.util.List;

@ApiStatus.Experimental
public class ModuleEventHandler {
    private final List<ModuleEventListener> listeners = new ArrayList<>();

    public void addListener(ModuleEventListener toAdd) {
        listeners.add(toAdd);
    }

    @ApiStatus.Internal
    public void runApplyEvent(Entity entity) {
        for (ModuleEventListener ml : listeners) {
            ml.onApply(entity);
        }
    }

    @ApiStatus.Internal
    public void runRemoveEvent(Entity entity) {
        for (ModuleEventListener ml : listeners) {
            ml.onRemove(entity);
        }
    }

    @ApiStatus.Internal
    public void runBreakEvent() {
        for (ModuleEventListener ml : listeners) {
            ml.onBreak();
        }
    }
}

File: src/main/kotlin/com/github/rochedo098/droids/api/events/ModuleEventHandler.kt
package com.github.rochedo098.droids.api.events

import net.minecraft.entity.Entity
import org.jetbrains.annotations.ApiStatus

class ModuleEventHandler {
    private val listeners: MutableList<ModuleEventListener> = ArrayList()

    fun addListener(toAdd: ModuleEventListener) { listeners.add(toAdd) }

    @ApiStatus.Internal
    fun runApplyEvent(entity: Entity?) {
        for (ls in listeners) {
            ls.onApply(entity)
        }
    }

    @ApiStatus.Internal
    fun runRemoveEvent(entity: Entity?) {
        for (ls in listeners) {
            ls.onRemove(entity)
        }
    }

    @ApiStatus.Internal
    fun runBreakEvent() {
        for (ls in listeners) {
            ls.onBreak()
        }
    }
}